### PR TITLE
[symbolizer] Support symbol+offset lookup

### PR DESCRIPTION
**Pull request by spavloff**
GNU addr2line supports lookup by symbol name in addition to the existing address lookup. llvm-symbolizer starting from e144ae54dcb96838a6176fd9eef21028935ccd4f supports lookup by symbol name. This change extends this lookup with possibility to specify optional offset.

Now the address for which source information is searched for can be specified with offset:

    llvm-symbolize --obj=abc.so "SYMBOL func_22+0x12"

It decreases the gap in features of llvm-symbolizer and GNU addr2line. This lookup now is supported for code only.

Migrated from: https://reviews.llvm.org/D139859.

### CODE REVIEW

File: llvm/test/tools/llvm-symbolizer/symbol-search.test [hunk 1]
```diff
@@ -34,6 +34,40 @@ RUN: llvm-symbolizer --obj=%p/Inputs/symbols.so func_01 func_02 | FileCheck --ch
...
+SYMLINE2: /tmp/dbginfo{{[/\]+}}symbols.part1.cpp:14
+
+# Offset can be specified with various bases.
+RUN: llvm-symbolizer --obj=%p/Inputs/symbols.so func_01+0x0C | FileCheck --check-prefix=SYMLINE1 %s
+RUN: llvm-addr2line -e %p/Inputs/symbols.so func_01+0x0C | FileCheck --check-prefix=SYMLINE1 %s
+
+RUN: llvm-symbolizer --obj=%p/Inputs/symbols.so func_01+014 | FileCheck --check-prefix=SYMLINE1 %s
+RUN: llvm-addr2line -e %p/Inputs/symbols.so func_01+014 | FileCheck --check-prefix=SYMLINE1 %s
+
+# If '+' is not followed by a number, it is a part of symbol, not an offset separator.
```
Comment (jh7370) [inline llvm/test/tools/llvm-symbolizer/symbol-search.test:57]:
```suggestion
# If '+' is not followed by a number, it is a part of the symbol name, not an offset separator.
```

File: llvm/test/tools/llvm-symbolizer/symbol-search.test [hunk 2]
```diff
@@ -34,6 +34,40 @@ RUN: llvm-symbolizer --obj=%p/Inputs/symbols.so func_01 func_02 | FileCheck --ch
...
+RUN: llvm-symbolizer --obj=%p/Inputs/symbols.so func_01+ | FileCheck --check-prefix=NONEXISTENT %s
+RUN: llvm-addr2line --obj=%p/Inputs/symbols.so func_01+ | FileCheck --check-prefix=NONEXISTENT %s
+
+RUN: llvm-symbolizer --obj=%p/Inputs/symbols.so func_01+abc | FileCheck --check-prefix=NONEXISTENT %s
+RUN: llvm-addr2line --obj=%p/Inputs/symbols.so func_01+abc | FileCheck --check-prefix=NONEXISTENT %s
+
+RUN: llvm-symbolizer --obj=%p/Inputs/symbols.so func_01+0A | FileCheck --check-prefix=NONEXISTENT %s
+RUN: llvm-addr2line --obj=%p/Inputs/symbols.so func_01+0A | FileCheck --check-prefix=NONEXISTENT %s
+
+# If '+' is not preceded by a symbol, it is a part of symbol, not an offset separator.
```
Comment (jh7370) [inline llvm/test/tools/llvm-symbolizer/symbol-search.test:67]:
```suggestion
# If '+' is not preceded by a symbol, it is a part of a symbol name, not an offset separator.
```

**Review by jh7370 — APPROVED** [review]
LGTM, with nits.

I asked in the Phabricator review whether GNU allows negative offsets. If it does, that functionality can be added in a follow-up patch.

File: llvm/test/tools/llvm-symbolizer/symbol-search.test (hunk 2 above)
Comment (spavloff) [reply llvm/test/tools/llvm-symbolizer/symbol-search.test]: Thanks!
> I asked in the Phabricator review whether GNU allows negative offsets. If it does, that functionality can be added in a follow-up patch.

No, only + is allowed for offsets: https://sourceware.org/git?p=binutils-gdb.git;a=blob;f=binutils/addr2line.c;h=fd2ac80dacd2fd14b353a7acf55aeb6f8d826288;hb=fd2ac80dacd2fd14b353a7acf55aeb6f8d826288#l252

File: llvm/test/tools/llvm-symbolizer/symbol-search.test [hunk 3]
```diff
@@ -54,7 +54,7 @@ RUN: llvm-addr2line -e %p/Inputs/symbols.so func_01+0x0C | FileCheck --check-pre
 RUN: llvm-symbolizer --obj=%p/Inputs/symbols.so func_01+014 | FileCheck --check-prefix=SYMLINE1 %s
 RUN: llvm-addr2line -e %p/Inputs/symbols.so func_01+014 | FileCheck --check-prefix=SYMLINE1 %s
 
-# If '+' is not followed by a number, it is a part of symbol, not an offset separator.
+# If '+' is not followed by a number, it is a part of the symbol name, not an offset separator.
```
Comment (jh7370) [inline llvm/test/tools/llvm-symbolizer/symbol-search.test:57]: Sorry, my mistake, this should be as suggested.
```suggestion
# If '+' is not followed by a number, it is part of the symbol name, not an offset separator.
```

File: llvm/test/tools/llvm-symbolizer/symbol-search.test [hunk 4]
```diff
@@ -64,7 +64,7 @@ RUN: llvm-addr2line --obj=%p/Inputs/symbols.so func_01+abc | FileCheck --check-p
 RUN: llvm-symbolizer --obj=%p/Inputs/symbols.so func_01+0A | FileCheck --check-prefix=NONEXISTENT %s
 RUN: llvm-addr2line --obj=%p/Inputs/symbols.so func_01+0A | FileCheck --check-prefix=NONEXISTENT %s
 
-# If '+' is not preceded by a symbol, it is a part of symbol, not an offset separator.
+# If '+' is not preceded by a symbol, it is a part of a symbol name, not an offset separator.
```
Comment (jh7370) [inline llvm/test/tools/llvm-symbolizer/symbol-search.test:67]:
```suggestion
# If '+' is not preceded by a symbol, it is part of a symbol name, not an offset separator.
```

**Pull request merged**